### PR TITLE
P20-585 Fix localizing Progression names with dots

### PR DESCRIPTION
**Pull request by mgc1194**
Some progression names were not showing their translations in CSC Marine Ecosystems despite of having translations available for several languages. Progressions are each of the levels students progress through within a lesson:
Screenshot 2024-01-12 at 14 31 13

During the `CourseContent::SyncIn`, progression names are pulled from each `Unit.script_levels`, and stored in `progressions.yml` using the name as key and value. 
```
progression_strings[script_level.progression] = script_level.progression
```
These names are given by curriculum writers and are not designed to be used as keys. The root cause of the "missing" translations is the usage of dots ` . ` in some string keys because our I18n rails library uses dots to indicate hierarchy:
`I18n.t("data.progressions.key")`
```
data:
  progressions:
    key: String
 ```

Fortunately, our[ i18n backend](https://github.com/code-dot-org/code-dot-org/blob/e5488d8173654f82360e8fec21482c85a917915f/lib/cdo/i18n_backend.rb#L22-L25) has an option that avoids this issue by setting the default separator to a character that is not in the key:
```
def self.get_smart_translate_options(locale, key, options = ::I18n::EMPTY_HASH)
        options = options.dup
        options.delete(:smart)

        # If I18n.t was called with an explicit scope and without specifying a
        # separator, try to find a separator character that isn't contained within
        # any of the individual scope or key values, to prevent a situation in
        # which periods in a key name can be mistakenly interpreted as separators
        if options.key?(:scope) && !options.key?(:separator)
          combined_key = [key] + options.fetch(:scope, [])
          options[:separator] = get_valid_separator(combined_key.join)
        end

        options
      end
```
This PR:
- Adds a `smart: true` option when localizing progressions in `script_level.rb`

## Links

- jira ticket: [P-585](https://codedotorg.atlassian.net/browse/P20-585)

## Testing story

Testing links: http://localhost-studio.code.org:3000/s/csc-ecosystems-2023

Screenshot 2024-01-12 at 00 52 08

## Follow-up work

Investigate if this issue is affecting other areas of the code base.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: bin/i18n/i18n_script_utils.rb [hunk 1]
```diff
@@ -336,6 +336,14 @@ def self.to_dashboard_i18n_data(locale, type, i18n_data)
     {locale => {'data' => {type => i18n_data}}}
   end
 
+  # Removes dots ( . ) from a given string.
+  # We frequently use strigns with dots as keys in our i18n data,
+  # as the come from course and curriculum content.
```
Comment (wilkie) [inline bin/i18n/i18n_script_utils.rb]: 'strings' typo, 'as they come' perhaps.

**Review by wilkie — APPROVED** [review]
This makes sense. It does seem to change the keys as seen by crowdin since the replacement happens on the sync-in. If that's ok, this is a good patch for these keys.

This very specifically seems to target one area. Is there any reasonable way we could detect these keys in other places? Is there any other place off-hand that could potentially fall into this issue as well?

**Comment by mgc1194** [comment]
> This makes sense. It does seem to change the keys as seen by crowdin since the replacement happens on the sync-in. If that's ok, this is a good patch for these keys.
> 
> This very specifically seems to target one area. Is there any reasonable way we could detect these keys in other places? Is there any other place off-hand that could potentially fall into this issue as well?

I agree that we don't know if this is an issue in other areas. It is a potential risk for sure, that is why I decided to create a method in the `I18n_sync_utils`, so we can apply the same patch wherever we encounter this issue.
I will investigate more, but we can always patch as bugs get reported.

File: dashboard/app/models/script_level.rb [hunk 2]
```diff
@@ -334,7 +334,10 @@ def summarize(include_prev_next = true, for_edit: false, user_id: nil)
 
       if progression
         summary[:progression] = progression
-        localized_progression_name = I18n.t("data.progressions.#{progression}", default: progression)
+        localized_progression_name = I18n.t(
+          "data.progressions.#{progression.gsub(I18n.default_separator, '')}",
+          default: progression
+        )
```
Comment (artem-vavilov) [inline dashboard/app/models/script_level.rb]: It seems we already have a solution for this (`smart: true`):

https://github.com/code-dot-org/code-dot-org/blob/54a55a398c3508cab5e4e55da1a47f7471cc5d7a/lib/cdo/i18n_backend.rb#L11-L13

```bash
[development] dashboard > I18n.t('If... then... Conditional Statements', scope: %i[data progressions], sm
art: true, locale: :'es-ES')
=> "Si... entonces... Declaraciones condicionales"
```

```suggestion
        localized_progression_name = I18n.t(
          progression,
          scope: %i[data progressions],
          default: progression,
          smart: true
        )
```

**Review by artem-vavilov — APPROVED** [review]
Nice work!

**Pull request merged**